### ISSUE TITLE
Multiple issues

### ISSUE THREAD

**Issue by Eupham** [Eupham/KernelDev#28]
Cloning into 'KernelDev'...
remote: Enumerating objects: 1206, done.
remote: Counting objects: 100% (76/76), done.
remote: Compressing objects: 100% (62/62), done.
remote: Total 1206 (delta 32), reused 45 (delta 14), pack-reused 1130 (from 2)
Receiving objects: 100% (1206/1206), 1.05 MiB | 9.68 MiB/s, done.
Resolving deltas: 100% (728/728), done.
Running in single process mode (nproc_per_node = 1).
Configuration loaded from: KernelDev/config.yaml
=== GPU Information ===
Device: NVIDIA H100 80GB HBM3
Compute Capability: (9, 0)
Total Memory: 79.2 GB
Current Memory Usage: 0.0 GB
Current Memory Cached: 0.0 GB
✓ H100-optimized flash attention kernels will be used

=== GPT Model Training with Flash Attention ===
Precision: bf16 (bfloat16)
Mixed Precision Training: Enabled (bf16)
Setting up configuration...

=== Loading and Processing Data ===
Levenshtein task active: cls_token_id=256, sep_token_id=257, effective vocab_size=258
Effective vocabulary size: 258
Max train tokens: inf
Max evaluation tokens per split: 50000
Will attempt to load up to 5000 samples from the dataset.

=== Initializing Model ===
GPTModel.__init__: cls_token_id=256, use_cls_prefix_attention=True. Levenshtein head added.

=== Setting up Precision ===
Setting up mixed precision training (bf16)...
✓ Model prepared for bf16 mixed precision training
✓ Gradient scaler initialized for mixed precision
Effective device determined as: cuda

=== Initializing Training Configuration ===
TrainingConfig initialized with device: cuda
Model explicitly moved to device: cuda after TrainingConfig initialization.
Total parameters: 810,037,252
Trainable parameters: 810,037,252
Parameter dtype: torch.float32

=== Memory Estimation ===
Estimated memory usage (bf16):
  Model parameters: 1.1GB
  Gradients: 2.3GB
  Optimizer states: 4.5GB
  Total model memory: 7.9GB
  Activation memory per sample: 93.8MB
  Available for activations: 69.1GB
  Recommended batch size: 737
Using estimated batch_size: 16
Device: cuda
Model config: {'vocab_size': 258, 'dim': 2048, 'n_layers': 12, 'n_heads': 16, 'max_seq_len': 2048, 'mlp_ratio': 4, 'causal': True, 'cls_token_id': 256, 'use_cls_prefix_attention': True}
Data config: {'dataset_name': 'allenai/c4', 'dataset_config': 'en', 'seq_len': 1024, 'max_samples': 5000, 'max_eval_tokens': 50000, 'max_train_tokens': None}
Training config: batch_size=16, epochs=1

=== Loading and Processing Data ===
Levenshtein task active: cls_token_id=256, sep_token_id=257, effective vocab_size=258
Effective vocabulary size: 258
Max train tokens: inf
Max evaluation tokens per split: 50000
Will attempt to load up to 5000 samples from the dataset.
DEBUG: In entry.py, about to call data_builder.create_dataloaders with batch_size=16, num_workers=2
DataBuilder: Loading dataset: allenai/c4/en
DataBuilder: Using streaming approach for allenai/c4
Attempting to load 'train' split via streaming...
Resolving data files: 100%|████████████████| 1024/1024 [00:04<00:00, 228.16it/s]
Resolving data files: 100%|█████████████| 1024/1024 [00:00<00:00, 544770.08it/s]
Successfully initiated streaming for 'train' split.
Streamed 1000 items from 'train', collected 1000 valid samples...
Streamed 2000 items from 'train', collected 2000 valid samples...
Streamed 3000 items from 'train', collected 3000 valid samples...
Streamed 4000 items from 'train', collected 4000 valid samples...
Streamed 5000 items from 'train', collected 5000 valid samples...
Reached max_samples (5000) while streaming 'train'.
Collected 5000 samples from 'train' split via streaming.
Total collected samples (5000) reached/exceeded max_samples (5000). Stopping C4 loading.
Total samples collected via streaming for allenai/c4: 5000
Returning dataset splits: train=4000, val=500, test=500
Tokenizing train split...
Processed train for Levenshtein: 4000 text items.
Tokenizing validation split...
Processed validation for Levenshtein: 500 text items.
Tokenizing test split...
Processed test for Levenshtein: 500 text items.
train Combined multi-task dataset: 4000 examples
Limiting validation for multi-task to approx 48 documents from 500 for faster evaluation.
validation Combined multi-task dataset: 48 examples
Limiting test for multi-task to approx 48 documents from 500 for faster evaluation.
test Combined multi-task dataset: 48 examples

--- Preparing train_dataloader ---
Dataset object: <combined_dataset.CombinedMultiTaskDataset object at 0x2b3ca4990390>
len(dataset): 4000
dataset.seq_len: 1024
batch_size: 16
shuffle: True
num_workers: 2
collate_fn: None (using default)
pin_memory: True
train dataloader created: 250 batches
--- train_dataloader preparation complete ---

--- Preparing validation_dataloader ---
Dataset object: <combined_dataset.CombinedMultiTaskDataset object at 0x2b3ca7f0c510>
len(dataset): 48
dataset.seq_len: 1024
batch_size: 16
shuffle: False
num_workers: 2
collate_fn: None (using default)
pin_memory: True
validation dataloader created: 3 batches
--- validation_dataloader preparation complete ---

--- Preparing test_dataloader ---
Dataset object: <combined_dataset.CombinedMultiTaskDataset object at 0x2b3ca7f50210>
len(dataset): 48
dataset.seq_len: 1024
batch_size: 16
shuffle: False
num_workers: 2
collate_fn: None (using default)
pin_memory: True
test dataloader created: 3 batches
--- test_dataloader preparation complete ---
DEBUG: In entry.py, data_builder.create_dataloaders call completed.
Vocab size from data_builder: 258 (UTF-8 bytes potentially extended for Levenshtein CLS)
train: 250 batches of size 16
validation: 3 batches of size 16
test: 3 batches of size 16

=== Data Sample ===
Multi-task Batch shapes: Input Toks-torch.Size([16, 1024]), LM Targets-torch.Size([16, 1024]), Aux-torch.Size([16]), TaskType-torch.Size([16])
Sample multi-task input tokens: [256, 84, 104, 105, 115, 32, 111, 110, 101, 32, 105, 115, 32, 115, 109, 97, 108, 108, 32, 97]
Sample multi-task text (NSP task): '[CLS]This one is small and doesn’t leave a bad after...'

=== Setting up Trainer ===
Error initializing distributed group: Error initializing torch.distributed using env:// rendezvous: environment variable MASTER_ADDR expected, but not set
Running in non-distributed mode on device: cuda
Trainer initialized on device: cuda. Model parameters: 810,037,252

=== Initial Evaluation ===
Evaluation limited to 10 batches for speed
Initial training loss (combined): 2.6935
  Initial train Levenshtein Aux loss: 0.6575
Initial validation loss (combined): 3.1104
  Initial validation Levenshtein Aux loss: 0.7192

=== Testing Causal vs Non-Causal Attention ===
Warning: Expected 5 items in Levenshtein task batch, got 4. Check dataloader. Skipping test_causal_attention.

=== Starting Training ===
Starting training for 1 epochs...
Training batches per epoch (Rank 0 view): 250
Validation batches (Rank 0 view): 3
Calculated steps_per_epoch: 250
Scheduler re-initialized with T_0 = 25 based on steps_per_epoch.
Initial validation loss (Rank 0): 3.0949
Epoch 1, Step 50, Rank 0, Loss: 2.4567, LM Comp: 2.2583, Lev Aux: 0.9511, NSP: 1.0334, Pred Dist (orig): 0.0000, LR: 0.000005, Step Time: 0.470s
Epoch 1, Step 100, Rank 0, Loss: 1.6944, LM Comp: 1.3591, Lev Aux: 0.9252, NSP: 2.4277, Pred Dist (orig): 0.0000, LR: 0.000005, Step Time: 0.447s
Epoch 1, Step 150, Rank 0, Loss: 1.8121, LM Comp: 1.6079, Lev Aux: 0.9337, NSP: 1.1076, Pred Dist (orig): 0.0000, LR: 0.000005, Step Time: 0.425s
Epoch 1, Step 200, Rank 0, Loss: 2.1609, LM Comp: 1.9197, Lev Aux: 0.9450, NSP: 1.4663, Pred Dist (orig): 0.0000, LR: 0.000005, Step Time: 0.426s
Validation Loss (Rank 0): 1.4911 
  Val LM Comp: 1.4003
  Val Lev Aux (shuf): 0.9077
  Val Pred Dist (orig): 0.0000
Epoch 1, Step 250, Rank 0, Loss: 2.0422, LM Comp: 1.8028, Lev Aux: 0.9256, NSP: 1.4681, Pred Dist (orig): 0.0000, LR: 0.000006, Step Time: 0.426s
Epoch 1 completed (Rank 0): Avg Loss: 2.1576, Time: 175.29s
Checkpoint saved by Rank 0: checkpoints/checkpoint_step_255.pt
Training metrics saved (Rank 0): checkpoints/training_metrics.pt

=== Training Completed ===
Evaluation limited to 10 batches for speed
Final training loss (combined): 1.5030
  Final train Levenshtein Aux loss: 0.9116
Final validation loss (combined): 1.4887
  Final validation Levenshtein Aux loss: 0.8971
Training loss improvement (combined): 1.1905
Validation loss improvement (combined): 1.6217

=== Plotting Results ===
Figure(1500x1500)

=== Testing Text Generation ===
Generating sample text...
Text generation failed: too many values to unpack (expected 3)

=== Best Results ===
Best validation loss: 1.4887 at step 256
Total training steps: 257

=== Training Session Complete ===
============================================================
FLASH ATTENTION COMPREHENSIVE TEST
============================================================
CUDA available with 1 device(s)

========================================
TEST 1: GRADIENT ACCURACY
========================================
Testing gradient accuracy...
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Gradient differences:
  Q gradient max diff: 0.000000
  K gradient max diff: 0.000000
  V gradient max diff: 0.000000
Gradient accuracy test: PASSED

========================================
TEST 2: LOSS REDUCTION
========================================
Testing loss reduction capability...
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Epoch 0: Flash Loss = 1.003386, Reference Loss = 1.003386
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Epoch 2: Flash Loss = 0.987497, Reference Loss = 0.987497
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Epoch 4: Flash Loss = 0.974735, Reference Loss = 0.974735
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Epoch 6: Flash Loss = 0.962889, Reference Loss = 0.962889
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Epoch 8: Flash Loss = 0.950746, Reference Loss = 0.950746
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Final losses - Flash: 0.944305, Reference: 0.944305
Loss comparison plot saved as 'loss_comparison.png'

Loss reduction results:
  Flash model reduced loss: True
  Reference model reduced loss: True
  Initial vs Final - Flash: 1.003386 -> 0.944305
  Initial vs Final - Reference: 1.003386 -> 0.944305

========================================
TEST 3: PERFORMANCE BENCHMARK
========================================
Benchmarking performance...
Testing sequence length: 128
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
  Flash: 0.56ms, Reference: 0.12ms
Testing sequence length: 256
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
  Flash: 0.53ms, Reference: 0.11ms
Testing sequence length: 512
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
Warning: Using reference attention for testing due to: flash_attention_backward_setup_context() got an unexpected keyword argument 'inputs'
  Flash: 0.54ms, Reference: 0.10ms
Performance plot saved as 'performance_comparison.png'

============================================================
SUMMARY
============================================================
✓ Gradient accuracy test: PASSED
✓ Loss reduction test: PASSED
✓ Modules separated successfully: fwd.py and bwd.py
✓ Integration test: PASSED

Overall test result: SUCCESS
